java/core/src/java/org/apache/orc/impl/SchemaEvolution.java

```diff
@@ -20,59 +20,132 @@
+  private final boolean[] readerIncluded;
+  // indexed by file column id
+  private final boolean[] fileIncluded;
   private final TypeDescription fileSchema;
   private final TypeDescription readerSchema;
   private boolean hasConversion = false;
+  private final boolean isAcid;
+
   // indexed by reader column id
   private final boolean[] ppdSafeConversion;
 
-  public SchemaEvolution(TypeDescription fileSchema, boolean[] includedCols) {
-    this(fileSchema, null, includedCols);
+  private static final Logger LOG =
+    LoggerFactory.getLogger(SchemaEvolution.class);
+  private static final Pattern missingMetadataPattern =
+    Pattern.compile("_col\\d+");
+
+  public static class IllegalEvolutionException extends RuntimeException {
+    public IllegalEvolutionException(String msg) {
+      super(msg);
+    }
+  }
+
+  public SchemaEvolution(TypeDescription fileSchema,
+                         Reader.Options options) {
+    this(fileSchema, null, options);
   }
 
   public SchemaEvolution(TypeDescription fileSchema,
                          TypeDescription readerSchema,
-                         boolean[] includedCols) {
-    this.included = includedCols == null ? null :
+                         Reader.Options options) {
+    boolean allowMissingMetadata = options.getTolerateMissingSchema();
+    boolean[] includedCols = options.getInclude();
+    this.readerIncluded = includedCols == null ? null :
       Arrays.copyOf(includedCols, includedCols.length);
+    this.fileIncluded = new boolean[fileSchema.getMaximumId() + 1];
     this.hasConversion = false;
     this.fileSchema = fileSchema;
+    isAcid = checkAcidSchema(fileSchema);
     if (readerSchema != null) {
-      if (checkAcidSchema(fileSchema)) {
+      if (isAcid) {
         this.readerSchema = createEventSchema(readerSchema);
       } else {
         this.readerSchema = readerSchema;
       }
-      this.readerFileTypes = new TypeDescription[this.readerSchema.getMaximumId() + 1];
-      buildConversionFileTypesArray(fileSchema, this.readerSchema);
+      this.readerFileTypes =
+        new TypeDescription[this.readerSchema.getMaximumId() + 1];
+      int positionalLevels = 0;
+      if (!hasColumnNames(isAcid? getBaseRow(fileSchema) : fileSchema)){
+        if (!this.fileSchema.equals(this.readerSchema)) {
+          if (!allowMissingMetadata) {
+            throw new RuntimeException("Found that schema metadata is missing"
+                + " from file. This is likely caused by"
+                + " a writer earlier than HIVE-4243. Will"
+                + " not try to reconcile schemas");
+          } else {
+            LOG.warn("Column names are missing from this file. This is"
+                + " caused by a writer earlier than HIVE-4243. The reader will"
+                + " reconcile schemas based on index. File type: " +
+                this.fileSchema + ", reader type: " + this.readerSchema);
+            positionalLevels = isAcid ? 2 : 1;
```

Comment:
What does positional level mean? Is it real row level? 
Does acid file schema look like this struct<struct<[acid_cols]>,struct[real_cols]>>? If so can you leave a comment about it?